### ISSUE TITLE
Support Azure Pipelines

### ISSUE THREAD

**Issue by jneira** [Mistuke/GhcChoco#9]
Hi, trying to reproduce https://gitlab.haskell.org/ghc/ghc/-/issues/18496 in azure i've found that
```
choco install -y --ignore-dependencies ghc --version=8.8.4 --force
```
did not install the specific ghc version, log:
```
Chocolatey v0.10.15
Installing the following packages:
ghc
By installing you accept licenses for the packages.

Progress: Downloading ghc 8.8.4... 9%
Progress: Downloading ghc 8.8.4... 28%
Progress: Downloading ghc 8.8.4... 48%
Progress: Downloading ghc 8.8.4... 67%
Progress: Downloading ghc 8.8.4... 87%
Progress: Downloading ghc 8.8.4... 100%

Chocolatey installed 0/0 packages. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
```
See https://dev.azure.com/jneira/e44fde34-db8b-435e-841c-4041b411213c/_apis/build/builds/816/logs/6 for example

**Comment by jneira** [comment]
I guess the bug in the lib settings about the linker location makes that version almost unusable in windows.
However there is a relatively easy workaround.

**Comment by Mistuke** [comment]
Hmm the log does not make much sense. Are you sure there's not an 8.10 installed already? Can you run the chocolatey command with `-d`?

**Comment by jneira** [comment]
yeah, there is a 8.10.1 already installed in the image, but i thought `--force` would install it anyway. In fact it works with ghc-8.8.3.

**Comment by Mistuke** [comment]
It looks like chocolatey is confused. Can you run it with `-d`?

**Comment by jneira** [comment]
I've ran it with -d:

```
2020-09-29T19:57:11.7193517Z Chocolatey v0.10.15
2020-09-29T19:57:11.7316465Z Chocolatey is running on Windows v 10.0.17763.0
2020-09-29T19:57:11.8054238Z Command line: "C:\ProgramData\Chocolatey\choco.exe" install -y -d --ignore-dependencies ghc --version=8.8.4 --force
2020-09-29T19:57:11.8065849Z Received arguments: install -y -d --ignore-dependencies ghc --version=8.8.4 --force
2020-09-29T19:57:33.2722988Z Performing validation checks.
2020-09-29T19:57:33.2797784Z Global Configuration Validation Checks:
2020-09-29T19:57:33.2819135Z  - Package Exit Code / Exit On Reboot = Checked
2020-09-29T19:57:33.2830043Z System State Validation Checks:
2020-09-29T19:57:33.2849230Z  Reboot Requirement Checks:
2020-09-29T19:57:33.2880126Z  - Pending Computer Rename = Checked
2020-09-29T19:57:33.2896663Z  - Pending Component Based Servicing = Checked
2020-09-29T19:57:33.2910110Z  - Pending Windows Auto Update = Checked
2020-09-29T19:57:33.2927508Z  - Pending File Rename Operations = Checked
2020-09-29T19:57:33.2943760Z  - Pending Windows Package Installer = Checked
2020-09-29T19:57:33.2967587Z  - Pending Windows Package Installer SysWow64 = Checked
2020-09-29T19:57:33.3067337Z The source 'https://chocolatey.org/api/v2/' evaluated to a 'normal' source type
2020-09-29T19:57:33.3087469Z 
2020-09-29T19:57:33.3089052Z NOTE: Hiding sensitive configuration data! Please double and triple
2020-09-29T19:57:33.3090139Z  check to be sure no sensitive data is shown, especially if copying
2020-09-29T19:57:33.3090958Z  output to a gist for review.
2020-09-29T19:57:33.3219028Z Configuration: CommandName='install'|
2020-09-29T19:57:33.3220194Z CacheLocation='C:\Users\VssAdministrator\AppData\Local\Temp\chocolatey'|
2020-09-29T19:57:33.3220834Z 
2020-09-29T19:57:33.3221116Z ContainsLegacyPackageInstalls='True'|
2020-09-29T19:57:33.3221680Z CommandExecutionTimeoutSeconds='2700'|WebRequestTimeoutSeconds='30'|
2020-09-29T19:57:33.3223050Z Sources='https://chocolatey.org/api/v2/'|SourceType='normal'|
2020-09-29T19:57:33.3223740Z Debug='True'|Verbose='False'|Trace='False'|Force='True'|Noop='False'|
2020-09-29T19:57:33.3227398Z HelpRequested='False'|UnsuccessfulParsing='False'|RegularOutput='True'|
2020-09-29T19:57:33.3228482Z QuietOutput='False'|PromptForConfirmation='False'|AcceptLicense='True'|
2020-09-29T19:57:33.3229679Z AllowUnofficialBuild='False'|Input='ghc'|Version='8.8.4'|
2020-09-29T19:57:33.3230663Z AllVersions='False'|SkipPackageInstallProvider='False'|
2020-09-29T19:57:33.3231482Z PackageNames='ghc'|Prerelease='False'|ForceX86='False'|
2020-09-29T19:57:33.3232398Z OverrideArguments='False'|NotSilent='False'|
2020-09-29T19:57:33.3234043Z ApplyPackageParametersToDependencies='False'|
2020-09-29T19:57:33.3235106Z ApplyInstallArgumentsToDependencies='False'|IgnoreDependencies='True'|
2020-09-29T19:57:33.3236092Z AllowMultipleVersions='False'|AllowDowngrade='False'|
2020-09-29T19:57:33.3236967Z ForceDependencies='False'|Information.PlatformType='Windows'|
2020-09-29T19:57:33.3239428Z Information.PlatformVersion='10.0.17763.0'|
2020-09-29T19:57:33.3240306Z Information.PlatformName='Windows Server 2016'|
2020-09-29T19:57:33.3241084Z Information.ChocolateyVersion='0.10.15.0'|
2020-09-29T19:57:33.3241959Z Information.ChocolateyProductVersion='0.10.15'|
2020-09-29T19:57:33.3243025Z Information.FullName='choco, Version=0.10.15.0, Culture=neutral, PublicKeyToken=79d02ea9cad655eb'|
2020-09-29T19:57:33.3243623Z 
2020-09-29T19:57:33.3244417Z Information.Is64BitOperatingSystem='True'|
2020-09-29T19:57:33.3245262Z Information.Is64BitProcess='True'|Information.IsInteractive='True'|
2020-09-29T19:57:33.3246136Z Information.UserName='VssAdministrator'|
2020-09-29T19:57:33.3247016Z Information.UserDomainName='WIN-3SAEL3SB0AG'|
2020-09-29T19:57:33.3248119Z Information.IsUserAdministrator='True'|
2020-09-29T19:57:33.3249487Z Information.IsUserSystemAccount='False'|
2020-09-29T19:57:33.3250302Z Information.IsUserRemoteDesktop='False'|
2020-09-29T19:57:33.3251165Z Information.IsUserRemote='True'|
2020-09-29T19:57:33.3252018Z Information.IsProcessElevated='True'|
2020-09-29T19:57:33.3252857Z Information.IsLicensedVersion='False'|Information.LicenseType='Foss'|
2020-09-29T19:57:33.3253845Z Information.CurrentDirectory='D:\a\1\s'|
2020-09-29T19:57:33.3255105Z Features.AutoUninstaller='True'|
2020-09-29T19:57:33.3255887Z Features.ChecksumFiles='True'|Features.AllowEmptyChecksums='False'|
2020-09-29T19:57:33.3256761Z Features.AllowEmptyChecksumsSecure='True'|
2020-09-29T19:57:33.3257576Z Features.FailOnAutoUninstaller='False'|
2020-09-29T19:57:33.3258493Z Features.FailOnStandardError='False'|Features.UsePowerShellHost='True'|
2020-09-29T19:57:33.3259488Z Features.LogEnvironmentValues='False'|Features.LogWithoutColor='False'|
2020-09-29T19:57:33.3260222Z Features.VirusCheck='False'|
2020-09-29T19:57:33.3261152Z Features.FailOnInvalidOrMissingLicense='False'|
2020-09-29T19:57:33.3262046Z Features.IgnoreInvalidOptionsSwitches='True'|
2020-09-29T19:57:33.3262811Z Features.UsePackageExitCodes='True'|
2020-09-29T19:57:33.3263666Z Features.UseEnhancedExitCodes='False'|
2020-09-29T19:57:33.3266435Z Features.UseFipsCompliantChecksums='False'|
2020-09-29T19:57:33.3267300Z Features.ShowNonElevatedWarnings='True'|
2020-09-29T19:57:33.3268154Z Features.ShowDownloadProgress='True'|
2020-09-29T19:57:33.3268918Z Features.StopOnFirstPackageFailure='False'|
2020-09-29T19:57:33.3269791Z Features.UseRememberedArgumentsForUpgrades='False'|
2020-09-29T19:57:33.3270766Z Features.IgnoreUnfoundPackagesOnUpgradeOutdated='False'|
2020-09-29T19:57:33.3271549Z Features.SkipPackageUpgradesWhenNotInstalled='False'|
2020-09-29T19:57:33.3272441Z Features.RemovePackageInformationOnUninstall='False'|
2020-09-29T19:57:33.3273211Z Features.ExitOnRebootDetected='False'|
2020-09-29T19:57:33.3274174Z Features.LogValidationResultsOnWarnings='True'|
2020-09-29T19:57:33.3275308Z Features.UsePackageRepositoryOptimizations='True'|
2020-09-29T19:57:33.3276101Z Features.ScriptsCheckLastExitCode='False'|
2020-09-29T19:57:33.3277003Z ListCommand.LocalOnly='False'|
2020-09-29T19:57:33.3277959Z ListCommand.IdOnly='False'|ListCommand.IncludeRegistryPrograms='False'|
2020-09-29T19:57:33.3278933Z ListCommand.PageSize='25'|ListCommand.Exact='False'|
2020-09-29T19:57:33.3279867Z ListCommand.ByIdOnly='False'|ListCommand.ByTagOnly='False'|
2020-09-29T19:57:33.3309179Z ListCommand.IdStartsWith='False'|ListCommand.OrderByPopularity='False'|
2020-09-29T19:57:33.3310140Z ListCommand.ApprovedOnly='False'|
2020-09-29T19:57:33.3311010Z ListCommand.DownloadCacheAvailable='False'|
2020-09-29T19:57:33.3311804Z ListCommand.NotBroken='False'|
2020-09-29T19:57:33.3312646Z ListCommand.IncludeVersionOverrides='False'|
2020-09-29T19:57:33.3313565Z UpgradeCommand.FailOnUnfound='False'|
2020-09-29T19:57:33.3316476Z UpgradeCommand.FailOnNotInstalled='False'|
2020-09-29T19:57:33.3317362Z UpgradeCommand.NotifyOnlyAvailableUpgrades='False'|
2020-09-29T19:57:33.3318162Z UpgradeCommand.ExcludePrerelease='False'|
2020-09-29T19:57:33.3319630Z NewCommand.AutomaticPackage='False'|
2020-09-29T19:57:33.3320629Z NewCommand.UseOriginalTemplate='False'|SourceCommand.Command='unknown'|
2020-09-29T19:57:33.3321455Z SourceCommand.Priority='0'|SourceCommand.BypassProxy='False'|
2020-09-29T19:57:33.3322332Z SourceCommand.AllowSelfService='False'|
2020-09-29T19:57:33.3323196Z SourceCommand.VisibleToAdminsOnly='False'|
2020-09-29T19:57:33.3325837Z FeatureCommand.Command='unknown'|ConfigCommand.Command='unknown'|
2020-09-29T19:57:33.3327616Z ApiKeyCommand.Remove='False'|PinCommand.Command='unknown'|
2020-09-29T19:57:33.3328782Z OutdatedCommand.IgnorePinned='False'|Proxy.BypassOnLocal='True'|
2020-09-29T19:57:33.3329598Z _ Chocolatey:ChocolateyInstallCommand - Normal Run Mode _
2020-09-29T19:57:33.3356150Z Installing the following packages:
2020-09-29T19:57:33.3364628Z ghc
2020-09-29T19:57:33.3371614Z By installing you accept licenses for the packages.
2020-09-29T19:57:35.1933660Z 
2020-09-29T19:57:35.1935548Z Progress: Downloading ghc 8.8.4... 9%
2020-09-29T19:57:35.1936421Z Progress: Downloading ghc 8.8.4... 28%
2020-09-29T19:57:35.1938209Z Progress: Downloading ghc 8.8.4... 48%
2020-09-29T19:57:35.1939314Z Progress: Downloading ghc 8.8.4... 67%
2020-09-29T19:57:35.1940136Z Progress: Downloading ghc 8.8.4... 87%
2020-09-29T19:57:35.1940958Z Progress: Downloading ghc 8.8.4... 100%
2020-09-29T19:57:35.2557766Z 
2020-09-29T19:57:35.2559570Z Chocolatey installed 0/0 packages. 
2020-09-29T19:57:35.2560556Z  See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
2020-09-29T19:57:35.3331640Z Exiting with 0
```

https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=825&view=results

**Comment by Mistuke** [comment]
Hmm according to that log the install finishes fine

```
2020-09-20 22:26:55,317 2280 [INFO ] -   Software installed to 'C:\ProgramData\chocolatey\lib\ghc.8.8.4\tools'
2020-09-20 22:26:55,317 2280 [DEBUG] - Attempting to delete file "C:\Users\packer\AppData\Local\NuGet\Cache\ghc.8.8.4.nupkg".
2020-09-20 22:26:55,333 2280 [WARN ] - 
Chocolatey installed 1/1 packages.
```
and configured correctly https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=825&view=logs&j=65acf690-ca6a-5c23-fc03-0de6fa53d8e8&t=2b4f35cc-7eb6-5c2e-f08c-8f37ebd439e1&l=41931

```
2020-09-20 22:26:34,714 2280 [INFO ] - C:\ProgramData\chocolatey\lib\ghc.8.8.4\tools
2020-09-20 22:26:35,012 2280 [DEBUG] - Running Install-ChocolateyPath -pathToInstall 'C:\ProgramData\chocolatey\lib\ghc.8.8.4\tools\ghc-8.8.4\bin' 
2020-09-20 22:26:35,012 2280 [DEBUG] - Running Update-SessionEnvironment 
2020-09-20 22:26:35,012 2280 [INFO ] - VERBOSE: Refreshing environment variables from the registry.
2020-09-20 22:26:35,153 2280 [INFO ] - PATH environment variable does not have C:\ProgramData\chocolatey\lib\ghc.8.8.4\tools\ghc-8.8.4\bin in it. Adding...
2020-09-20 22:26:35,168 2280 [INFO ] - VERBOSE: Choosing not to expand environment names
2020-09-20 22:26:35,184 2280 [DEBUG] - Running Set-EnvironmentVariable -Name 'Path' -Value 'C:\Rust\.cargo\bin;%USERPROFILE%\AppData\Local\Microsoft\WindowsApps;C:\Users\packer\AppData\Roaming\npm;C:\Program Files (x86)\GitHub CLI\;C:\Users\packer\AppData\Roaming\Composer\vendor\bin;%USERPROFILE%\.dotnet\tools;C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\bin;C:\Users\packer\AppData\Roaming\cabal\bin;C:\ProgramData\chocolatey\lib\ghc.8.6.5\tools\ghc-8.6.5\bin;C:\ProgramData\chocolatey\lib\ghc.8.8.4\tools\ghc-8.8.4\bin;' -Scope 'User' 
2020-09-20 22:26:35,199 2280 [DEBUG] - Registry type for Path is/will be ExpandString
2020-09-20 22:26:35,215 2280 [DEBUG] -
```

But shows that immediately after finishing it's told to install `8.10.2`.

```
2020-09-20 22:26:59,264 3156 [INFO ] - [NuGet] Installing 'ghc 8.10.2'.
2020-09-20 22:26:59,279 3156 [DEBUG] - [NuGet] Added file 'chocolateyInstall.ps1' to folder 'ghc.8.10.2\tools'.
2020-09-20 22:26:59,279 3156 [DEBUG] - [NuGet] Added file 'ghc.8.10.2.nupkg' to folder 'ghc.8.10.2'.
2020-09-20 22:26:59,546 3156 [DEBUG] - [NuGet] Added file 'ghc.8.10.2.nuspec' to folder 'ghc.8.10.2'
```

Though a call to 

```
2020-09-20 22:26:57,465 3156 [DEBUG] - Command line: "C:\ProgramData\Chocolatey\choco.exe" install ghc -y --version 8.10.2 -m
2020-09-20 22:26:57,465 3156 [DEBUG] - Received arguments: install ghc -y --version 8.10.2 -m
```

https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=825&view=logs&j=65acf690-ca6a-5c23-fc03-0de6fa53d8e8&t=2b4f35cc-7eb6-5c2e-f08c-8f37ebd439e1&l=61285

https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=825&view=logs&j=65acf690-ca6a-5c23-fc03-0de6fa53d8e8&t=2b4f35cc-7eb6-5c2e-f08c-8f37ebd439e1&l=61406

The problem is that `-m` likely doesn't work correctly on github actions because they don't allow very fine control of `PATH`.
To get it to work the package emits a directive https://github.com/Mistuke/GhcChoco/blob/master/8.8.4/ghc/tools/chocolateyInstall.ps1#L52

but it seems the last one to emit this clobbers the existing ones?

Argg looks like for Azure devops/pipelines you have yet a different method... https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/authoring/commands.md

Is this github actions of azure pipelines?

**Comment by jneira** [comment]
It is azure pipelines.
Just in case, the log you are commenting is the result of make a `cat "C:\ProgramData\chocolatey\logs\chocolatey.log"` just after trying to (re)install ghc-8.8.4. So the log is about the original install when setting up the image itself.

Sorry for not metioning it, were you aware of the double timestamp in [the raw log](https://dev.azure.com/jneira/e44fde34-db8b-435e-841c-4041b411213c/_apis/build/builds/825/logs/7)? it seems the web view is hidding the original timestamp (it was 2020-09-29) and only show the one from chocolatey.log (2020-09-20)

```
2020-09-29T19:57:45.5521954Z 2020-09-20 22:26:55,317 2280 [INFO ] -  The install of ghc was successful.
2020-09-29T19:57:45.5523065Z 2020-09-20 22:26:55,317 2280 [INFO ] -   Software installed to 'C:\ProgramData\chocolatey\lib\ghc.8.8.4\tools'
2020-09-29T19:57:45.5525041Z 2020-09-20 22:26:55,317 2280 [DEBUG] - Attempting to delete file "C:\Users\packer\AppData\Local\NuGet\Cache\ghc.8.8.4.nupkg".
2020-09-29T19:57:45.5525970Z 2020-09-20 22:26:55,333 2280 [WARN ] - 
2020-09-29T19:57:45.5526768Z Chocolatey installed 1/1 packages. 
```
So the original install of ghc-8.8.4 was succesful but overwritten by ghc-8.10.2, but my new attempt to install with `-f` failed.
I've ran another pipeline without `chocolatey.log` to avoid confusions:
https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=833&view=logs&j=65acf690-ca6a-5c23-fc03-0de6fa53d8e8&t=2b4f35cc-7eb6-5c2e-f08c-8f37ebd439e1

There is the [my unique try to install it with `-f`](https://github.com/jneira/floskell-test/blob/ghc-8.8.4/.azure/windows.yml#L8), that ends with
```
2020-09-30T14:39:58.3385815Z Chocolatey installed 0/0 packages. 
```

**Comment by Mistuke** [comment]
> Sorry for not mentioning it, were you aware of the double timestamp in the raw log? it seems the web view is hidding the original timestamp (it was 2020-09-29) and only show the one from chocolatey.log (2020-09-20)

Hmm no I wasn't.. The output for the `8.8.4` install isn't verbose enough, but this is starting to feel like a caching issue, can you run it with `-dv` instead of just `-d` to up the verbosity?

```
      choco install -y -d --ignore-dependencies ghc --version=8.8.4 --force
      choco install -y --ignore-dependencies ghc --version=8.10.1 --force
      choco install -y --ignore-dependencies ghc --version=8.10.2
```

Why not just

```
      choco install -y -d --ignore-dependencies ghc --version=8.8.4 -m
      choco install -y --ignore-dependencies ghc --version=8.10.1 -m
      choco install -y --ignore-dependencies ghc --version=8.10.2 -m
```

as your goal seems to be to have all 3 installed at the same time?
With force I really only expect 8.10.2 to be left.

**Comment by jneira** [comment]
Thanks for taking care of this.
In my first tries i saw in the log of a `choco install -y --ignore-dependencies ghc --version=8.10.1`:
```
Use --allow-downgrade or --force to attempt to install older versions, or use side by side to allow multiple versions.
```
So i used `--force` to install the older ones. In fact using it with ghc-8.8.3 and ghc-8.10.1, it installs those versions in `/chocolatey/lib/ghc/tools` but no for ghc-8.8.4. Reading the docs about `-m` i see that is just the `use side by side` 🤦‍♂️ 

I am just start a execution with `-dv`and `-m`:  https://dev.azure.com/jneira/haskell-language-server/_build/results?buildId=834&view=results

**Comment by Mistuke** [comment]
So I see `[NuGet] 'ghc 8.8.4' already installed.`

Which is why the log is showing 

```
2020-09-29T19:57:35.2559570Z Chocolatey installed 0/0 packages. 
```

I'm terribly confused, why is it already installed? is it installed in the base image as well?

**Comment by Mistuke** [comment]
I'll double check the installer package tomorrow, but it should be the same as 8.10..

What do you see if you `ls $GHCS_PATH` ? (remove the `-dv` to reduce output :)) and also `echo $PATH` after it's set?

from a quick reading of https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch I'm not sure export works..

**Comment by jneira** [comment]
I.ve just found the script that installs them: https://github.com/actions/virtual-environments/blob/main/images/win/scripts/Installers/Install-Haskell.ps1

It installs ghc-8.6.5, ghc-8.8.4 and ghc-8.10.2, but in `%APPDATA%\chocolatey\lib\ghc.8.10.2`, `%APPDATA%\chocolatey\lib\ghc.8.8.4` and  `%APPDATA%\chocolatey\lib\ghc.8.6.5`.
I expected them will be all in `%APPDATA%\chocolatey\lib\ghc\tools`. In fact ghc-8.8.3 and ghc-8.10.1 were being installed in that location and in my local machine all ghcs are in that location too.
 
Now i've learnt that `-m` installs packages in those locations.

So it seems you can force to install 8.8.3 and 8.10.1 as they are not already installed.

Finally all is reduced to my inexperience with choco, sorry for the inconveniences.

**Comment by Mistuke** [comment]
Ahh yeah that makes sense. Yeah with `-m` the paths get versioned so they don't conflict.

Still you shouldn't need to manually modify the path, the reason for that is that I wasn't aware azure pipeline had it's own way of setting paths (all these CI services seem to have their own ways).

I'll add the Azure pipelines runes to the next releases.

**Comment by Mistuke** [comment]
The latest version of the packages also support global installs, which should also work(based on the comment on https://github.com/actions/virtual-environments/blob/main/images/win/scripts/Installers/Install-Haskell.ps1) See https://chocolatey.org/packages/ghc

**Comment by Mistuke** [comment]
Fixed in 33ee2d0c4fbd40e0ef0c2bc7da81beb597421698 for 8.10.2.1 and newer.  8.10.2.1 also post processes the settings file to fix linker issue.
